Compilation issue

```
Scanning dependencies of target libydk_yang_untar
[  1%] Unpacking openconfig@0.1.3.tar.gz
[  1%] Built target libydk_yang_untar
Scanning dependencies of target ydk_openconfig
[  3%] Building CXX object CMakeFiles/ydk_openconfig.dir/ydk/models/openconfig/generated_entity_lookup.cpp.o
[  5%] Building CXX object CMakeFiles/ydk_openconfig.dir/ydk/models/openconfig/iana_if_type.cpp.o
In file included from /Users/xiaoqinz/Cisco/devnet/gen-api/cpp/openconfig-bundle/ydk/models/openconfig/iana_if_type.cpp:7:
/Users/xiaoqinz/Cisco/devnet/gen-api/cpp/openconfig-bundle/ydk/models/openconfig/iana_if_type.hpp:15:34: error:
      use of undeclared identifier 'ietf_interfaces'; did you mean 'ietf::ietf_interfaces'?
class IanaInterfaceType : public ietf_interfaces::InterfaceType, virtual ydk::Identity
                                 ^~~~~~~~~~~~~~~
                                 ietf::ietf_interfaces
/usr/local/include/ydk_ietf/ietf_interfaces.hpp:12:11: note: 'ietf::ietf_interfaces' declared here
namespace ietf_interfaces {
          ^
1 error generated.
make[2]: *** [CMakeFiles/ydk_openconfig.dir/ydk/models/openconfig/iana_if_type.cpp.o] Error 1
make[1]: *** [CMakeFiles/ydk_openconfig.dir/all] Error 2
make: *** [all] Error 2
```

Change package CMakeLists.txt template or bundle name for headers might help with this issue.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -18,7 +18,7 @@ The [backward incompatible](http://ydk.cisco.com/py/docs/backward_compatibility.
    * `YFilter` has replaced the functionality of the `READ` and `DELETE` objects
    * When using logging, the suggested level for users of YDK is `INFO` as `DEBUG` provides highly detailed logging suitable for developers working on YDK
    * The type names of `enumerations` and `identities` no longer have `Enum` or `Identity` in their names. For example, the  identity `InterfaceTypeIdentity` in `ydk.models.ietf.ietf_interfaces` is now renamed to just `InterfaceType`
-
+   * The ``is_config()`` method is no longer available for the YDK model APIs.    This may be added back in a future release.
 
 #### C++
  * Renamed some of the methods in path API to be verbs like `DataNode::get_schema()` instead of nouns like `DataNode::schema()` ([#498](https://github.com/CiscoDevNet/ydk-gen/pull/498))

--- a/README.md
+++ b/README.md
@@ -53,7 +53,7 @@ Of course, many other libraries are used as an integral part of ydk-gen and its 
 The output of ydk-gen is either a core package, that defines services and providers, or a module bundle, consisting of APIs based on YANG models. Each module bundle is generated using a bundle profile and the ydk-gen tool. Developers can either use pre-packaged generated bundles (e.g. [ydk-py](http://cs.co/ydk-py)), or they can define their own bundle, consisting of a set of YANG models, using a bundle profile (e.g. [```ietf_0_1_1.json```](profiles/bundles/ietf_0_1_1.json)). This gives a developer the ability to customize the scope of their bundle based on their requirements.
 
 
-# Note
+# Backward compatibility
  Please see [this page](http://ydk.cisco.com/py/docs/backward_compatibility.html) for details on some backward incompatible changes introduced as part of 0.6.0 release
 
 

--- a/sdk/cpp/tests/test_xml_subtree.cpp
+++ b/sdk/cpp/tests/test_xml_subtree.cpp
@@ -194,8 +194,8 @@ bgp->global->config->as = 65172;
 bgp->global->config->router_id = "1.2.3.4";
 
 auto afi_safi = make_unique<openconfig_bgp::Bgp::Global::AfiSafis::AfiSafi>();
-afi_safi->afi_safi_name = openconfig_bgp_types::L3Vpn_Ipv4_Unicast();
-afi_safi->config->afi_safi_name = openconfig_bgp_types::L3Vpn_Ipv4_Unicast();
+afi_safi->afi_safi_name = openconfig_bgp_types::L3VPN_IPV4_UNICAST();
+afi_safi->config->afi_safi_name = openconfig_bgp_types::L3VPN_IPV4_UNICAST();
 afi_safi->config->enabled = true;
 afi_safi->parent = bgp->global->afi_safis.get();
 bgp->global->afi_safis->afi_safi.push_back(move(afi_safi));

--- a/sdk/python/CHANGES.md
+++ b/sdk/python/CHANGES.md
@@ -1,7 +1,6 @@
 ### 2017-08-01 version 0.6.0
 
 #### Python
-* **NOTE:** Please see [this page](http://ydk.cisco.com/py/docs/backward_compatibility.html) for details on some backward incompatible changes introduced as part of this release
 * Introduced new YDK python [`core`](https://github.com/CiscoDevNet/ydk-py/tree/master/core) package using [pybind11](https://github.com/pybind/pybind11) to wrap around YDK C++ [`core`](https://github.com/CiscoDevNet/ydk-cpp/tree/master/core) ([#507](https://github.com/CiscoDevNet/ydk-gen/pull/507))
   * Introduced `ydk.path` module consisting of APIs to read, manipulate and write YANG data using XPath-like expressions
   * Updated YDK services and providers to internally use the path API
@@ -20,6 +19,17 @@
 * Updated [`cisco-ios-xr`](https://github.com/CiscoDevNet/ydk-gen/blob/master/profiles/bundles/cisco-ios-xr_6_2_2.json) to support Cisco IOS XR 6.2.2 release
 * Updated [`cisco-ios-xe`](https://github.com/CiscoDevNet/ydk-gen/blob/master/profiles/bundles/cisco-ios-xe_16_6_1.json) to support Cisco IOS XE 16.6.1 release
 * Also updated [`openconfig`](https://github.com/CiscoDevNet/ydk-gen/blob/master/profiles/bundles/openconfig_0_1_3.json) and [`ietf`](https://github.com/CiscoDevNet/ydk-gen/blob/master/profiles/bundles/ietf_0_1_3.json) bundles
+
+#### Note on backward compatibility
+The [backward incompatible](http://ydk.cisco.com/py/docs/backward_compatibility.html) changes introduced with `0.6.0` release include:
+
+ * **Installation:** When installing `YDK-Py`, there is a new system requirement which needs to be installed. This is the `libydk` library, which is available on the DevHub website for various OS platforms. Please refer to the [system requirements](https://github.com/CiscoDevNet/ydk-py/blob/master/README.rst#system-requirements) for details.
+* **Windows support:** From release ``0.6.0`` onwards, YDK is no longer supported on the Windows platform. We plan to add back support for this platform in the future.
+* **API changes:** Please refer to the [developer guide](http://ydk.cisco.com/py/docs/developer_guide.html) and [API guide](http://ydk.cisco.com/py/docs/api_guide.html) for details about APIs and how to use them.
+   * `NetconfServiceProvider` no longer has the `close()` method. There is no need to explicitly close the provider as it will be automatically cleaned up when the object goes out of scope.
+   * `YFilter` has replaced the functionality of the `READ` and `DELETE` objects
+   * When using logging, the suggested level for users of YDK is `INFO` as `DEBUG` provides highly detailed logging suitable for dvelopers working on YDK
+   * The type names of `enumerations` and `identities` no longer have `Enum` or `Identity` in their names. For example, the  identity `InterfaceTypeIdentity` in `ydk.models.ietf.ietf_interfaces` is now renamed to just `InterfaceType`.
 
 ### 2017-06-06 version 0.5.5
 

--- a/sdk/python/README.rst
+++ b/sdk/python/README.rst
@@ -12,13 +12,18 @@ Overview
 
 The YANG Development Kit (YDK) is a Software Development Kit that provides API's that are modeled in YANG. The main goal of YDK is to reduce the learning curve of YANG data models by expressing the model semantics in an API and abstracting protocol/encoding details.  YDK is composed of a core package that defines services and providers, plus one or more module bundles that are based on YANG models.
 
+Backward Compatibility
+----------------------
+Please see `the release notes <https://github.com/CiscoDevNet/ydk-py/releases/tag/0.6.0>`_ and `documentation <http://ydk.cisco.com/py/docs/backward_compatibility.html>`_ for details on some backward incompatible changes introduced as part of the 0.6.0 release.
+
+
 System Requirements
 -------------------
 Linux
 ~~~~~
 Ubuntu (Debian-based) - The following packages must be present in your system before installing YDK-Py::
 
-   $ sudo apt-get install gdebi-core python3-dev python-dev
+   $ sudo apt-get install gdebi-core python3-dev python-dev libtool-bin
    $ wget https://devhub.cisco.com/artifactory/debian-ydk/0.6.0/libydk_0.6.0-1_amd64.deb
    $ sudo gdebi libydk_0.6.0-1_amd64.deb
 

--- a/sdk/python/core/docsgen/guides/backward_compatibility.rst
+++ b/sdk/python/core/docsgen/guides/backward_compatibility.rst
@@ -24,3 +24,4 @@ API Changes
 2. :py:class:`YFilter<ydk.filters.YFilter>` has replaced the functionality of the ``READ`` and ``DELETE`` objects
 3. When using YDK's :ref:`howto-logging`, the suggested level to be used is ``INFO``
 4. The type names of ``enumerations`` and ``identities`` no longer have ``Enum`` or ``Identity`` in their names. For example, the  identity ``InterfaceTypeIdentity`` in ``ydk.models.ietf.ietf_interfaces`` is now renamed to just :py:class:`InterfaceType<ydk.models.ietf.ietf_interfaces.InterfaceType>`.
+5. The ``is_config()`` method is no longer available for the YDK model APIs. This may be added back in a future release.

--- a/sdk/python/core/samples/_config_builder.py
+++ b/sdk/python/core/samples/_config_builder.py
@@ -25,13 +25,13 @@ def _get_bgp_config():
     bgp_cfg.global_.config.as_ = 65001
 
     ipv4_afsf = bgp_cfg.global_.afi_safis.AfiSafi()
-    ipv4_afsf.afi_safi_name = openconfig_bgp_types.Ipv4Unicast()
-    ipv4_afsf.config.afi_safi_name = openconfig_bgp_types.Ipv4Unicast()
+    ipv4_afsf.afi_safi_name = openconfig_bgp_types.IPV4_UNICAST()
+    ipv4_afsf.config.afi_safi_name = openconfig_bgp_types.IPV4_UNICAST()
     ipv4_afsf.config.enabled = True
 
     ipv6_afsf = bgp_cfg.global_.afi_safis.AfiSafi()
-    ipv6_afsf.afi_safi_name = openconfig_bgp_types.Ipv6Unicast()
-    ipv6_afsf.config.afi_safi_name = openconfig_bgp_types.Ipv6Unicast()
+    ipv6_afsf.afi_safi_name = openconfig_bgp_types.IPV6_UNICAST()
+    ipv6_afsf.config.afi_safi_name = openconfig_bgp_types.IPV6_UNICAST()
     ipv6_afsf.config.enabled = True
 
     bgp_cfg.global_.afi_safis.afi_safi.append(ipv4_afsf)
@@ -45,9 +45,9 @@ def _get_bgp_config():
     nbr_ipv4.config.peer_as = 65002
 
     nbr_ipv4_afsf = nbr_ipv4.afi_safis.AfiSafi()
-    nbr_ipv4_afsf.afi_safi_name = openconfig_bgp_types.Ipv4Unicast()
+    nbr_ipv4_afsf.afi_safi_name = openconfig_bgp_types.IPV4_UNICAST()
     nbr_ipv4_afsf.config.peer_as = 65002
-    nbr_ipv4_afsf.config.afi_safi_name = openconfig_bgp_types.Ipv4Unicast()
+    nbr_ipv4_afsf.config.afi_safi_name = openconfig_bgp_types.IPV4_UNICAST()
     nbr_ipv4_afsf.config.enabled = True
 
     # Create afi-safi policy instances
@@ -68,11 +68,11 @@ def _get_routing_cfg():
     routing_policy.policy_definitions.policy_definition.append(pass_all_policy_defn)
     pass_all_policy_defn.parent = routing_policy.policy_definitions
 
-    comm_set = RoutingPolicy.DefinedSets.BgpDefinedSets.CommunitySets.CommunitySet()
-    comm_set.community_set_name = 'testing'
-    comm_set.community_member.append("65172:16001")
-    comm_set.community_member.append("65172:16032")
-    routing_policy.defined_sets.bgp_defined_sets.community_sets.community_set.append(comm_set)
+#    RoutingPolicy.DefinedSets.BgpDefinedSets.CommunitySets.CommunitySet()
+#    comm_set.community_set_name = 'testing'
+#comm_set.config.community_member.append("65172:16001")
+#comm_set.config.community_member.append("65172:16032")
+#    routing_policy.defined_sets.bgp_defined_sets.community_sets.community_set.append(comm_set)
     return routing_policy
 
 

--- a/sdk/python/core/setup.py
+++ b/sdk/python/core/setup.py
@@ -108,7 +108,7 @@ def get_python_library():
         library_prefix = sysconfig.get_config_var('PYTHONFRAMEWORKPREFIX')
         library_path = os.path.join(library_prefix, sysconfig.get_config_var('LDLIBRARY'))
 
-    assert os.path.exists(library_path), 'Could not find python library in {0}. Check your python installation'.format(library_path)
+#assert os.path.exists(library_path), 'Could not find python library in {0}. Check your python installation'.format(library_path)
     return library_path
 
 

--- a/sdk/python/core/ydk/types/py_types.py
+++ b/sdk/python/core/ydk/types/py_types.py
@@ -22,11 +22,13 @@
         - Entity
 """
 from ydk.ext.types import ChildrenMap
+from ydk.ext.types import Enum as _Enum
 from ydk.ext.types import YLeaf as _YLeaf
 from ydk.ext.types import YLeafList as _YLeafList
 from ydk.ext.types import Entity as _Entity
 from ydk.filters import YFilter as _YFilter
 from ydk.errors import YPYModelError as _YPYModelError
+from ydk.errors.error_handler import handle_type_error as _handle_type_error
 
 
 class YList(list):
@@ -159,3 +161,28 @@ class Entity(_Entity):
         if not isinstance(value, obj.__class__):
             raise _YPYModelError("Invalid value '{!s}' in '{}'"
                                  .format(value, obj))
+
+    def _perform_setattr(self, clazz, leaf_names, name, value):
+        self._check_monkey_patching_error(name, value)
+        with _handle_type_error():
+            if name in self.__dict__ and isinstance(self.__dict__[name], YList):
+                raise _YPYModelError("Attempt to assign value of '{}' to YList ldata. "
+                                    "Please use list append or extend method."
+                                    .format(value))
+            if isinstance(value, _Enum.YLeaf):
+                value = value.name
+            if name in leaf_names and name in self.__dict__:
+                if isinstance(value, _YLeaf):
+                    self.__dict__[name].set(value.get())
+                elif isinstance(value, _YLeafList):
+                    super(clazz, self).__setattr__(name, value)
+                else:
+                    self.__dict__[name].set(value)
+            else:
+                if hasattr(value, "parent") and name != "parent":
+                    if hasattr(value, "is_presence_container") and value.is_presence_container:
+                        value.parent = self
+                    elif value.parent is None and value.yang_name in self._children_yang_names:
+                        value.parent = self
+                super(clazz, self).__setattr__(name, value)
+

--- a/sdk/python/core/ydk/types/py_types.py
+++ b/sdk/python/core/ydk/types/py_types.py
@@ -175,7 +175,7 @@ class Entity(_Entity):
                 if isinstance(value, _YLeaf):
                     self.__dict__[name].set(value.get())
                 elif isinstance(value, _YLeafList):
-                    super(clazz, self).__setattr__(name, value)
+                    super(Entity, self).__setattr__(name, value)
                 else:
                     self.__dict__[name].set(value)
             else:
@@ -184,5 +184,5 @@ class Entity(_Entity):
                         value.parent = self
                     elif value.parent is None and value.yang_name in self._children_yang_names:
                         value.parent = self
-                super(clazz, self).__setattr__(name, value)
+                super(Entity, self).__setattr__(name, value)
 

--- a/ydkgen/__init__.py
+++ b/ydkgen/__init__.py
@@ -122,9 +122,9 @@ class YdkGenerator(object):
         _set_original_bundle_name_for_packages(all_bundles, packages, curr_bundle)
         gen_api_root = self._init_bundle_directories(packages=packages, bundle=curr_bundle)
 
-        generated_files = self._print_packages(packages, gen_api_root, curr_bundle)
+        yang_models = _create_models_archive(curr_bundle, gen_api_root)
 
-        yang_models = self._create_models_archive(curr_bundle, gen_api_root)
+        generated_files = self._print_packages(packages, gen_api_root, curr_bundle)
 
         if self.language == 'cpp':
             _modify_cpp_cmake(gen_api_root, curr_bundle.name, curr_bundle.str_version, 
@@ -179,8 +179,9 @@ class YdkGenerator(object):
         """
         global classes_per_source_file
         factory = printer_factory.PrinterFactory()
+        bundle_packages = _filter_bundle_from_packages(pkgs, bundle)
         ydk_printer = factory.get_printer(self.language)(output_dir, bundle, self.generate_tests, self.sort_clazz)
-        generated_files = ydk_printer.emit(pkgs, classes_per_source_file)
+        generated_files = ydk_printer.emit(bundle_packages, classes_per_source_file)
         return generated_files
 
     # Initialize generated API directory ######################################
@@ -255,30 +256,56 @@ class YdkGenerator(object):
         logger.debug('Copying %s to %s' % (target_dir, gen_api_root))
         dir_util.copy_tree(target_dir, gen_api_root)
 
-    def _create_models_archive(self, bundle, target_dir):
-        '''
-        Creates yang models archive as part of bundle package.
-        Args:
-            source_dir (str): Directory where models are located
-            bundle_qualified_name (st): Bundle name with version
-            target_dir (str): Directory where archive is to be created
-        '''
-        global YDK_YANG_MODEL
-        assert isinstance(bundle, bundle_resolver.Bundle)
-        tar_file = '{}.tar.gz'.format(bundle.fqn)
-        tar_file_path = os.path.join(target_dir, tar_file)
-        ydk_yang = os.path.join(target_dir, YDK_YANG_MODEL)
-        if os.path.isfile(ydk_yang):
-            shutil.copy(ydk_yang, bundle.resolved_models_dir)
-        yang_models = []
-        with tarfile.open(tar_file_path, 'w:gz') as tar:
-            for yang_model in os.listdir(bundle.resolved_models_dir):
-                if yang_model.endswith(".yang"):
-                    yang_models.append(yang_model)
-                    yang_model_path = os.path.join(bundle.resolved_models_dir, yang_model)
-                    tar.add(yang_model_path, arcname=os.path.basename(yang_model_path))
-        logger.debug('\nCreated models archive: {}'.format(tar_file_path))
-        return yang_models
+
+def _filter_bundle_from_packages(pkgs, bundle):
+    bundle_packages = []
+    bundle_package_names = [x.name for x in bundle.models]
+    for package in pkgs:
+        if package.stmt.arg in bundle_package_names:
+            bundle_packages.append(package)
+    return bundle_packages
+
+
+def _get_yang_models_filenames_from_directory(path, prefix):
+    yang_models = []
+    for file in os.listdir(path):
+        file_path = os.path.join(path, file)
+        if file.endswith(".yang"):
+            yang_models.append(os.path.join(prefix, file))
+        elif os.path.islink(file_path):
+            yang_models.extend(_get_yang_models_filenames_from_directory(file_path, file))
+    return yang_models
+
+
+def _create_tar(resolved_models_dir, tar_file_path):
+    yang_models = _get_yang_models_filenames_from_directory(resolved_models_dir, '')
+    with tarfile.open(tar_file_path, 'w:gz') as tar:
+        for y in yang_models:
+            yang_model_path = os.path.join(resolved_models_dir, y)
+            tar.add(yang_model_path, arcname=os.path.basename(yang_model_path))
+    yang_models_base_names = [os.path.basename(file) for file in yang_models]
+    return yang_models_base_names
+
+
+def _create_models_archive(bundle, target_dir):
+    '''
+    Creates yang models archive as part of bundle package.
+    Args:
+        source_dir (str): Directory where models are located
+        bundle_qualified_name (st): Bundle name with version
+        target_dir (str): Directory where archive is to be created
+    '''
+    global YDK_YANG_MODEL
+    assert isinstance(bundle, bundle_resolver.Bundle)
+    tar_file = '{}.tar.gz'.format(bundle.fqn)
+    tar_file_path = os.path.join(target_dir, tar_file)
+    ydk_yang = os.path.join(target_dir, YDK_YANG_MODEL)
+    if os.path.isfile(ydk_yang):
+        shutil.copy(ydk_yang, bundle.resolved_models_dir)
+    yang_models = _create_tar(bundle.resolved_models_dir, tar_file_path)
+
+    logger.debug('\nCreated models archive: {}'.format(tar_file_path))
+    return yang_models
 
 
 def _set_original_bundle_name_for_packages(bundles, packages, curr_bundle):
@@ -433,4 +460,3 @@ def _check_description_file(description_file):
     if not os.path.isfile(description_file):
         logger.error('Path to description file is not valid.')
         raise YdkGenException('Path to description file is not valid.')
-

--- a/ydkgen/api_model.py
+++ b/ydkgen/api_model.py
@@ -875,10 +875,10 @@ def snake_case(input_text):
 
 
 def camel_case(input_text):
-    def _title(str):
-        if len(str) > 0  and str.startswith(str[0].upper()):
-            return str
-        return str.title()
+    def _title(s):
+        if len(s) > 0  and str.startswith(s[0].upper()):
+            return s
+        return s.title()
     return ''.join([_title(word) for word in input_text.split('-')])
 
 

--- a/ydkgen/api_model.py
+++ b/ydkgen/api_model.py
@@ -214,6 +214,7 @@ class NamedElement(Element):
 
     def fully_qualified_cpp_name(self):
         ''' get the C++ qualified name '''
+        pkg = get_top_pkg(self)
         names = []
         element = self
         while element is not None:
@@ -221,7 +222,7 @@ class NamedElement(Element):
                 element = element.owner
             names.append(element.name)
             element = element.owner
-        return '::'.join(reversed(names))
+        return pkg.bundle_name + '::' + '::'.join(reversed(names))
 
 
 class Package(NamedElement):
@@ -874,7 +875,11 @@ def snake_case(input_text):
 
 
 def camel_case(input_text):
-    return ''.join([word.title() for word in input_text.split('-')])
+    def _title(str):
+        if len(str) > 0  and str.startswith(str[0].upper()):
+            return str
+        return str.title()
+    return ''.join([_title(word) for word in input_text.split('-')])
 
 
 def escape_name(name):

--- a/ydkgen/builder/_api_model_builder.py
+++ b/ydkgen/builder/_api_model_builder.py
@@ -78,6 +78,7 @@ class ApiModelBuilder(object):
         if enum_type_stmt is not None:
             enum_class = Enum(self.iskeyword)
             enum_class.stmt = enum_type_stmt
+            disambiguate_class_name_from_ancestors_and_siblings(self.language, enum_class, pe)
             enum_type_stmt.parent.i_enum = enum_class
             enum_type_stmt.i_enum = enum_class
             pe.owned_elements.append(enum_class)
@@ -194,6 +195,7 @@ class ApiModelBuilder(object):
             # we have to create the enum
             enum_class = Enum(self.iskeyword)
             enum_class.stmt = enum_type
+            disambiguate_class_name_from_ancestors_and_siblings(self.language, enum_class, parent_element)
             parent_element.owned_elements.append(enum_class)
             enum_class.owner = parent_element
             prop.property_type = enum_class
@@ -214,6 +216,7 @@ class ApiModelBuilder(object):
                     if contained_enum_type is not None and contained_enum_type == contained_type:
                         enum_class = Enum(self.iskeyword)
                         enum_class.stmt = contained_enum_type
+                        disambiguate_class_name_from_ancestors_and_siblings(self.language, enum_class, parent_element)
                         parent_element.owned_elements.append(enum_class)
                         enum_class.owner = parent_element
                         contained_enum_type.i_enum = enum_class
@@ -619,3 +622,13 @@ def snake_case(input_text):
     s = input_text.replace('-', '_')
     s = s.replace('.', '_')
     return s.lower()
+
+
+def disambiguate_class_name_from_ancestors_and_siblings(language, clazz, parent_element):
+    if language == 'cpp':
+        if name_matches_ancestor(clazz.name, parent_element):
+            clazz.name = clazz.name + '_'
+        for e in parent_element.owned_elements:
+            if e.name == clazz.name:
+                clazz.name = clazz.name + '_'
+

--- a/ydkgen/printer/python/class_docstring_printer.py
+++ b/ydkgen/printer/python/class_docstring_printer.py
@@ -47,6 +47,7 @@ class ClassDocstringPrinter(object):
         self._print_class_docstring_presence(clazz)
         self.ctx.writeln('"""')
         self.ctx.bline()
+        self.ctx.lvl_dec()
 
     def _print_class_docstring_text(self, clazz):
         class_docstring = get_class_docstring(clazz, 'py')

--- a/ydkgen/printer/python/class_inits_printer.py
+++ b/ydkgen/printer/python/class_inits_printer.py
@@ -177,59 +177,11 @@ class ClassSetAttrPrinter(object):
     def _print_class_setattr_header(self):
         self.ctx.writeln('def __setattr__(self, name, value):')
         self.ctx.lvl_inc()
-        self.ctx.writeln('self._check_monkey_patching_error(name, value)')
-        self.ctx.writeln('with _handle_type_error():')
-        self.ctx.lvl_inc()
 
     def _print_class_setattr_body(self, clazz, leafs):
-        leaf_names = [ '"%s"' % (leaf.name) for leaf in leafs ]
-        separator = ',\n%s%s' % (self.ctx.tab(), ' '*12)
-
-        self.ctx.writeln('if name in self.__dict__ and isinstance(self.__dict__[name], YList):')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('raise YPYModelError("Attempt to assign value of \'{}\' to YList ldata. "')
-        self.ctx.writeln('                    "Please use list append or extend method."')
-        self.ctx.writeln('                    .format(value))')
-        self.ctx.lvl_dec()
-        self.ctx.writeln('if isinstance(value, Enum.YLeaf):')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('value = value.name')
-        self.ctx.lvl_dec()
-        self.ctx.writeln('if name in (%s) and name in self.__dict__:' %
-            separator.join(leaf_names))
-        self.ctx.lvl_inc()
-        self.ctx.writeln('if isinstance(value, YLeaf):')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('self.__dict__[name].set(value.get())')
-        self.ctx.lvl_dec()
-        self.ctx.writeln('elif isinstance(value, YLeafList):')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('super(%s, self).__setattr__(name, value)' % clazz.qn())
-        self.ctx.lvl_dec()
-
-        self.ctx.writeln('else:')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('self.__dict__[name].set(value)')
-        self.ctx.lvl_dec()
-
-        self.ctx.lvl_dec()
-        self.ctx.writeln('else:')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('if hasattr(value, "parent") and name != "parent":')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('if hasattr(value, "is_presence_container") and value.is_presence_container:')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('value.parent = self')
-        self.ctx.lvl_dec()
-        self.ctx.writeln('elif value.parent is None and value.yang_name in self._children_yang_names:')
-        self.ctx.lvl_inc()
-        self.ctx.writeln('value.parent = self')
-        self.ctx.lvl_dec()
-        self.ctx.lvl_dec()
-        self.ctx.writeln('super(%s, self).__setattr__(name, value)' % clazz.qn())
-        self.ctx.lvl_dec()
+        leaf_names = ['%s' % (leaf.name) for leaf in leafs]
+        self.ctx.writeln('self._perform_setattr(%s, %s, name, value)'%(clazz.qn(), leaf_names))
 
     def _print_class_setattr_trailer(self):
-        self.ctx.lvl_dec()
         self.ctx.lvl_dec()
         self.ctx.bline()

--- a/ydkgen/printer/python/class_printer.py
+++ b/ydkgen/printer/python/class_printer.py
@@ -120,6 +120,7 @@ class ClassPrinter(FilePrinter):
         ClassDocstringPrinter(self.ctx).print_output(clazz)
 
     def _print_class_attributes(self, clazz):
+        self.ctx.lvl_inc()
         module = clazz.stmt
         if not clazz.stmt.keyword == 'module':
             module = clazz.stmt.i_module

--- a/ydkgen/printer/python/python_bindings_printer.py
+++ b/ydkgen/printer/python/python_bindings_printer.py
@@ -21,7 +21,7 @@
 from __future__ import print_function
 
 
-import os
+import os, shutil
 from distutils import dir_util
 
 from ydkgen.api_model import Bits, Class, Enum
@@ -59,7 +59,7 @@ class PythonBindingsPrinter(LanguageBindingsPrinter):
             self._print_python_rst_ydk_models()
 
         # YANG models
-        self._print_yang_files()
+        self._copy_yang_files()
         return ()
 
     def _print_modules(self):
@@ -162,10 +162,20 @@ class PythonBindingsPrinter(LanguageBindingsPrinter):
                         emit_nmsp_declare_init,
                         _EmitArgs(self.ypy_ctx, self.packages))
 
-    def _print_yang_files(self):
+    def _copy_yang_files_from_subdirectories(self, yang_files_dir):
+        subdirs = [os.path.join(yang_files_dir, o) for o in os.listdir(yang_files_dir) if os.path.isdir(os.path.join(yang_files_dir, o))]
+        for subdir in subdirs:
+            files = os.listdir(subdir)
+            for file in files:
+                file_path = os.path.join(subdir, file)
+                if os.path.isfile(file_path):
+                    shutil.copy(file_path, yang_files_dir)
+
+    def _copy_yang_files(self):
         yang_files_dir = os.path.sep.join([self.models_dir, '_yang'])
         os.mkdir(yang_files_dir)
         dir_util.copy_tree(self.bundle.resolved_models_dir, yang_files_dir)
+        self._copy_yang_files_from_subdirectories(yang_files_dir)
 
 
 def get_init_file_name(path):

--- a/ydkgen/printer/python/python_bindings_printer.py
+++ b/ydkgen/printer/python/python_bindings_printer.py
@@ -162,20 +162,21 @@ class PythonBindingsPrinter(LanguageBindingsPrinter):
                         emit_nmsp_declare_init,
                         _EmitArgs(self.ypy_ctx, self.packages))
 
-    def _copy_yang_files_from_subdirectories(self, yang_files_dir):
-        subdirs = [os.path.join(yang_files_dir, o) for o in os.listdir(yang_files_dir) if os.path.isdir(os.path.join(yang_files_dir, o))]
-        for subdir in subdirs:
-            files = os.listdir(subdir)
-            for file in files:
-                file_path = os.path.join(subdir, file)
-                if os.path.isfile(file_path):
-                    shutil.copy(file_path, yang_files_dir)
-
     def _copy_yang_files(self):
         yang_files_dir = os.path.sep.join([self.models_dir, '_yang'])
         os.mkdir(yang_files_dir)
         dir_util.copy_tree(self.bundle.resolved_models_dir, yang_files_dir)
-        self._copy_yang_files_from_subdirectories(yang_files_dir)
+        _copy_yang_files_from_subdirectories(yang_files_dir)
+
+
+def _copy_yang_files_from_subdirectories(yang_files_dir):
+    subdirs = [os.path.join(yang_files_dir, o) for o in os.listdir(yang_files_dir) if os.path.isdir(os.path.join(yang_files_dir, o))]
+    for subdir in subdirs:
+        files = os.listdir(subdir)
+        for file in files:
+            file_path = os.path.join(subdir, file)
+            if os.path.isfile(file_path):
+                shutil.copy(file_path, yang_files_dir)
 
 
 def get_init_file_name(path):

